### PR TITLE
Add ktlint support (without formatting) for kotlin filetype

### PR DESCRIPTION
**Pull request by francis36012**
Hey @w0rp, this adds ktlint support for kotlin  (#593). There is no fix functionality because I'm not sure how that should be implemented.

### CODE REVIEW

File: ale_linters/kotlin/ktlint.vim [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        if len(l:match) == 0
+            continue
+        endif
+
+        let l:file = l:match[1]
+        let l:line = l:match[2] + 0
+        let l:column = l:match[3] + 0
+        let l:text = l:match[4]
+
+        let l:buf_abspath = fnamemodify(l:file, ':p')
```
Comment (w0rp) [inline ale_linters/kotlin/ktlint.vim]: This variable is unused.

File: ale_linters/kotlin/ktlint.vim [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    let l:output = []
+
+    for l:line in a:lines
+        let l:match = matchlist(l:line, l:message_pattern)
+
+        if len(l:match) == 0
+            continue
+        endif
+
+        let l:file = l:match[1]
```
Comment (w0rp) [inline ale_linters/kotlin/ktlint.vim]: This variable is unused.

File: ale_linters/kotlin/ktlint.vim [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+function! ale_linters#kotlin#ktlint#Handle(buffer, lines) abort
+    let l:message_pattern = '^\(.*\):\([0-9]\+\):\([0-9]\+\):\s\+\(.*\)'
+    let l:output = []
+
+    for l:line in a:lines
+        let l:match = matchlist(l:line, l:message_pattern)
+
+        if len(l:match) == 0
+            continue
+        endif
```
Comment (w0rp) [inline ale_linters/kotlin/ktlint.vim]: You can use `for l:match in ale#util#GetMatches(a:lines, l:message_pattern)`.

File: ale_linters/kotlin/ktlint.vim [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        let l:text = l:match[4]
+
+        let l:buf_abspath = fnamemodify(l:file, ':p')
+        let l:type = l:text =~? 'not a valid kotlin file' ? 'E' : 'W'
+
+        call add(l:output, {
+            \ 'lnum': l:line,
+            \ 'col': l:column,
+            \ 'text': l:text,
+            \ 'type': l:type
```
Comment (w0rp) [inline ale_linters/kotlin/ktlint.vim]: Put your `\` line continuation markers at the start of your lines, like the other files.

File: ale_linters/kotlin/ktlint.vim [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+let g:ale_kotlin_ktlint_exec = get(g:, 'ale_kotlin_ktlint_exec', 'ktlint')
+let g:ale_kotlin_ktlint_rulesets = get(g:, 'ale_kotlin_ktlint_rulesets', [])
+let g:ale_kotlin_ktlint_format = get(g: , 'ale_kotlin_ktlint_format', 0)
+let g:ale_kotlin_ktlint_config_file = get(g:, 'ale_kotlin_ktlint_config_file', '.ale_kotlin_ktlint_config')
+let g:ale_kotlin_ktlint_enable_config = get(g:, 'ale_kotlin_ktlint_enable_config', 0)
+
+
+function! ale_linters#kotlin#ktlint#GetCommand(buffer) abort
+    if ale#Var(a:buffer, 'kotlin_ktlint_enable_config')
+        let l:config_file = ale#path#FindNearestFile(a:buffer, ale#Var(a:buffer, 'kotlin_ktlint_config_file'))
```
Comment (w0rp) [inline ale_linters/kotlin/ktlint.vim]: You can remove this. `g:ale_pattern_options` lets you configure different files based on regular expressions for paths, so we don't need to source other files to set up variables.

**Comment by w0rp** [comment]
For fixing files, look through the codebase for existing tools for fixing files with `:ALEFix`. Fixing and linting are done separately. You probably can't fix files with this linter yet for two reasons.

1. ALE doesn't yet support fixing and linting at the same time with the same command.
2. ALE doesn't yet support fixing files on disk.

File: ale_linters/kotlin/ktlint.vim (hunk 5 above)
Comment (francis36012) [inline ale_linters/kotlin/ktlint.vim]: After reading the documentation for `g:ale_patterns_options`, I'm wondering how one will configure options for a specific directory. That is, how will options intended for a specific project (or directory) be set without having to add the directory to `g:ale_pattern_options`?

Comment (w0rp) [inline ale_linters/kotlin/ktlint.vim]: You just add a pattern for the directory.

Comment (francis36012) [inline ale_linters/kotlin/ktlint.vim]: And this is set in the `vimrc`? BTW, will the IRC channel be a better place for this sort of discussion?

Comment (w0rp) [inline ale_linters/kotlin/ktlint.vim]: Yeah, I suppose so. I'll be going to bed soon, but I always lurk on there and can check the backlog whenever.

Comment (francis36012) [inline ale_linters/kotlin/ktlint.vim]: Hey @w0rp all changes have been made

File: ale_linters/kotlin/ktlint.vim [hunk 6]
```diff
@@ -0,0 +1,54 @@
+" Author: Francis Agyapong <francisagyapong2@gmail.com>
+" Description: Lint kotlin files using ktlint
+
+let g:ale_kotlin_ktlint_exec = get(g:, 'ale_kotlin_ktlint_exec', 'ktlint')
```
Comment (w0rp) [inline ale_linters/kotlin/ktlint.vim]: Sorry I didn't notice this earlier, but this option should be named `executable` so it matches the other option names.

If you like, you can also use `ale#Set` to configure variables, if you hate typing the same name twice. Look for examples in the codebase.

**Comment by w0rp** [comment]
This is looking pretty good. :+1:

For the benefit of others, we decided that we can add a feature for loading a configuration file, for any linters. I'll open an issue about it later.

**Comment by w0rp** [comment]
Cheers! :beers:

**Pull request merged**